packages/camera/camera_avfoundation/ios/camera_avfoundation/Sources/camera_avfoundation/CameraPlugin.swift

```diff
@@ -0,0 +1,528 @@
+      forName: UIDevice.orientationDidChangeNotification,
+      object: UIDevice.current,
+      queue: .main
+    ) { [weak self] notification in
+      self?.orientationChanged(notification)
+    }
+  }
+
+  public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
+    UIDevice.current.endGeneratingDeviceOrientationNotifications()
+  }
+
+  static func flutterErrorFromNSError(_ error: NSError) -> FlutterError {
+    return FlutterError(
+      code: "Error \(error.code)", message: error.localizedDescription, details: error.domain)
+  }
+
+  func orientationChanged(_ notification: Notification) {
+    guard let device = notification.object as? UIDevice else { return }
+    let orientation = device.orientation
+
+    if orientation == .faceUp || orientation == .faceDown {
+      // Do not change when oriented flat.
+      return
+    }
+
+    self.captureSessionQueue.async { [weak self] in
+      // `FLTCam.setDeviceOrientation` must be called on capture session queue.
+      self?.camera?.setDeviceOrientation(orientation)
+      // `CameraPlugin.sendDeviceOrientation` can be called on any queue.
+      self?.sendDeviceOrientation(orientation)
+    }
+  }
+
+  func sendDeviceOrientation(_ orientation: UIDeviceOrientation) {
+    DispatchQueue.main.async { [weak self] in
+      self?.globalEventAPI.deviceOrientationChangedOrientation(
+        FCPGetPigeonDeviceOrientationForOrientation(orientation)
+      ) { _ in
+        // Ignore errors; this is essentially a broadcast stream, and
+        // it's fine if the other end doesn't receive the message
+        // (e.g., if it doesn't currently have a listener set up).
+      }
+    }
+  }
+
+  public func availableCameras(
+    completion: @escaping ([FCPPlatformCameraDescription]?, FlutterError?) -> Void
+  ) {
+    captureSessionQueue.async { [weak self] in
+      guard let strongSelf = self else { return }
+
+      var discoveryDevices: [AVCaptureDevice.DeviceType] = [
+        .builtInWideAngleCamera,
+        .builtInTelephotoCamera,
+      ]
+
+      if #available(iOS 13.0, *) {
+        discoveryDevices.append(.builtInUltraWideCamera)
+      }
+
+      let devices = strongSelf.deviceDiscoverer.discoverySession(
+        withDeviceTypes: discoveryDevices,
+        mediaType: .video,
+        position: .unspecified)
+
+      var reply: [FCPPlatformCameraDescription] = []
+
+      for device in devices {
+        var lensFacing: FCPPlatformCameraLensDirection
+
+        switch device.position {
+        case .back:
+          lensFacing = .back
+        case .front:
+          lensFacing = .front
+        case .unspecified:
+          lensFacing = .external
+        @unknown default:
+          lensFacing = .external
+        }
+
+        let cameraDescription = FCPPlatformCameraDescription.make(
+          withName: device.uniqueID,
+          lensDirection: lensFacing
+        )
+        reply.append(cameraDescription)
+      }
+
+      completion(reply, nil)
```

Comment:
do we ever complete with error?

Reply:
No, but the completion type is required by the `FCPCameraApi` protocol so I don't think we can do anything about it